File: frontend/src/services/quiz.ts
import api from './apiClient';
import { CanonicalQuestionV1, QuizReportData } from '../types';
import { parseApi } from '../contracts/parse';
import { QuizReportDataSchema } from '../contracts/schemas';
import { mapQuizReportResponseToQuizReportData } from './mappers';


const makeLocalReportId = (): string => {
  if (typeof crypto !== 'undefined' && 'randomUUID' in crypto) {
    return crypto.randomUUID();
  }
  return `local_report_${Date.now()}_${Math.random().toString(36).slice(2)}`;
};

const saveLocalQuizReport = (data: any): string => {
  const id = data.id || makeLocalReportId();
  const report: QuizReportData = {
    id,
    student_name: data.studentName || 'Guest User',
    quiz_topic: data.topic || 'Practice Quiz',
    score: data.score || 0,
    total_questions: data.total || 0,
    correct_count: data.correct || 0,
    created_at: new Date().toISOString(),
    answers_snapshot: data.answers || [],
  };

  const stored = localStorage.getItem('quiz_reports');
  const reports = stored ? JSON.parse(stored) : [];
  const next = [report, ...reports.filter((r: any) => r.id !== id)].slice(0, 50);
  localStorage.setItem('quiz_reports', JSON.stringify(next));
  return id;
};

export interface Question {
  id: string;
  skill_id: number;
  section: string;
  interaction: string;
  stimulus?: string;
  prompt: string;
  choices?: string[];
  correct_response?: string[];
  cefr_target?: string;
  difficulty_score?: number;
  passage_id?: string;
  metadata?: {
    explanation?: string;
    pattern_tip?: string;
    hints?: string[];
    source?: string;
  };
  created_at: string;
}

export interface QuestionFilter {
  section?: string;
  skill_id?: number;
  cefr?: string;
  limit?: number;
}

export interface QuizResult {
  id: string;
  user_id: string;
  date: string;
  skill_id?: string;
  section: string;
  score: number;
  correct_count: number;
  total_questions: number;
  xp_earned: number;
  breakdown?: {
    correct: number;
    wrong: number;
    skipped: number;
  };
}

export interface SaveResultRequest {
  skill_id?: string;
  section: string;
  score: number;
  correct_count: number;
  total_questions: number;
}

export interface AdaptiveMetricsResponse {
  total_questions: number;
  correct_answers: number;
  accuracy_by_section: Record<string, { correct: number; total: number }>;
  accuracy_by_skill: Record<string, { correct: number; total: number }>;
  recent_accuracy: number[];
  average_response_time: number;
  last_updated: number;
  current_difficulty: string;
}

export interface RecordAnswerRequest {
  correct: boolean;
  section: string;
  skill_id: string;
  response_time_ms: number;
}

export interface ProgressResponse {
  total_quizzes: number;
  total_correct: number;
  total_xp: number;
  unique_skills: number;
  level: number;
}

export interface HistoryItem {
  id: string;
  date: string;
  section: string;
  skill_id?: string;
  score: number;
  correct_count: number;
  total_questions: number;
  xp_earned: number;
}

export const quizService = {
  async listQuestions(filter?: QuestionFilter): Promise<Question[]> {
    const params = new URLSearchParams();
    if (filter?.section) params.set('section', filter.section);
    if (filter?.skill_id) params.set('skill_id', filter.skill_id.toString());
    if (filter?.cefr) params.set('cefr', filter.cefr);
    if (filter?.limit) params.set('limit', filter.limit.toString());

    const query = params.toString();
    const response = await api.get<Question[]>(`/api/quiz/questions${query ? `?${query}` : ''}`);
    return response.data || [];
  },

  async simulation(section?: string): Promise<Question[]> {
    const query = section ? `?section=${section}` : '';
    const response = await api.get<Question[]>(`/api/quiz/simulation${query}`);
    return response.data || [];
  },

  async saveResult(data: SaveResultRequest): Promise<{ ok: boolean; id?: string; xp_earned?: number; next_difficulty_level?: string; error?: string }> {
    const response = await api.post<{ ok: boolean; id: string; xp_earned: number; next_difficulty_level?: string }>('/api/quiz/results', data);
    if (response.error) {
      return { ok: false, error: response.error.error };
    }
    if (response.data) {
      return {
        ok: response.data.ok,
        id: response.data.id,
        xp_earned: response.data.xp_earned,
        next_difficulty_level: response.data.next_difficulty_level,
      };
    }
    return { ok: false, error: 'Unknown error' };
  },

  async history(): Promise<HistoryItem[]> {
    const response = await api.get<HistoryItem[]>('/api/quiz/history');
    return response.data || [];
  },

  async progress(): Promise<ProgressResponse | null> {
    const response = await api.get<ProgressResponse>('/api/quiz/progress');
    return response.data || null;
  },

  async getAdaptiveMetrics(): Promise<AdaptiveMetricsResponse | null> {
    const response = await api.get<AdaptiveMetricsResponse>('/api/quiz/adaptive-metrics');
    return response.data || null;
  },

  async recordAnswer(data: RecordAnswerRequest): Promise<AdaptiveMetricsResponse | null> {
    const response = await api.post<AdaptiveMetricsResponse>('/api/quiz/record-answer', data);
    return response.data || null;
  },

  async getQuizReportById(id: string): Promise<QuizReportData | null> {
    const response = await api.get<QuizReportData>(`/api/quiz/reports/${id}`);
    if (response.data) {
      return parseApi(QuizReportDataSchema, mapQuizReportResponseToQuizReportData(response.data));
    }
    try {
      const stored = localStorage.getItem('quiz_reports');
      const reports = stored ? JSON.parse(stored) : [];
      return reports.find((r: any) => r.id === id) || null;
    } catch {
      return null;
    }
  },

  async saveQuizReport(data: any): Promise<string | null> {
    try {
      const response = await api.post<{ ok: boolean; id: string }>('/api/quiz/reports', {
        skill_id: data.skillId ? String(data.skillId) : undefined,
        section: data.section,
        student_name: data.studentName,
        quiz_topic: data.topic,
        score: data.score,
        total_questions: data.total,
        correct_count: data.correct,
        answers_snapshot: data.answers,
      });
      if (response.data?.id) {
        return response.data.id;
      }
      return saveLocalQuizReport(data);
    } catch {
      return saveLocalQuizReport(data);
    }
  }
};

export default quizService;

File: src/models/quiz.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use vil_orm_derive::VilEntity;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SectionAccuracy {
    pub correct: i64,
    pub total: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "user_performance_metrics")]
pub struct UserPerformanceMetrics {
    #[vil_entity(pk)]
    pub id: String,
    #[vil_entity(unique)]
    pub user_id: String,
    pub total_questions: i64,
    pub correct_answers: i64,
    pub accuracy_by_section: String, // JSON
    pub accuracy_by_skill: String,   // JSON
    pub recent_accuracy: String,     // JSON array
    pub average_response_time: f64,
    pub current_difficulty: String,
    pub last_updated: i64,
}

#[derive(Debug, Deserialize)]
pub struct RecordAnswerRequest {
    pub correct: bool,
    pub section: String,
    pub skill_id: String,
    pub response_time_ms: i64,
}

#[derive(Debug, Serialize)]
pub struct AdaptiveMetricsResponse {
    pub total_questions: i64,
    pub correct_answers: i64,
    pub accuracy_by_section: HashMap<String, SectionAccuracy>,
    pub accuracy_by_skill: HashMap<String, SectionAccuracy>,
    pub recent_accuracy: Vec<i64>,
    pub average_response_time: f64,
    pub last_updated: i64,
    pub current_difficulty: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "question_bank")]
pub struct Question {
    #[vil_entity(pk)]
    pub id: String,
    pub skill_id: i64,
    pub section: String,
    pub interaction: String,
    pub stimulus: Option<String>, // JSON
    pub prompt: String,
    pub choices: Option<String>,          // JSON
    pub correct_response: Option<String>, // JSON
    pub cefr_target: Option<String>,
    pub difficulty_score: Option<i64>,
    pub passage_id: Option<String>,
    pub metadata: Option<String>, // JSON
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct Passage {
    pub id: String,
    pub topic: Option<String>,
    pub content: String,
    pub source: Option<String>,
    pub difficulty: Option<String>,
    pub word_count: Option<i64>,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow, VilEntity)]
#[vil_entity(table = "quiz_results")]
pub struct QuizResult {
    #[vil_entity(pk)]
    pub id: String,
    pub user_id: String,
    pub date: String,
    pub skill_id: Option<String>,
    pub section: String,
    pub score: i64,
    pub correct_count: i64,
    pub total_questions: i64,
    pub xp_earned: i64,
    pub breakdown: Option<String>, // JSON
}

#[derive(Debug, Deserialize)]
pub struct SaveResultRequest {
    pub skill_id: Option<String>,
    pub section: String,
    pub score: i64,
    pub correct_count: i64,
    pub total_questions: i64,
}

#[derive(Debug, Deserialize)]
pub struct QuestionFilter {
    pub section: Option<String>,
    pub skill_id: Option<i64>,
    pub cefr: Option<String>,
    pub limit: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct QuizListParams {
    pub section: Option<String>,
    pub skill_id: Option<i64>,
    pub cefr: Option<String>,
    pub page: Option<i64>,
    pub limit: Option<i64>,
}

#[derive(Debug, Deserialize)]
pub struct QuizGenerateRequest {
    pub topic: String,
    pub section: String,
    pub count: Option<i64>,
    pub skill_id_override: Option<i64>,
    pub difficulty: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GeneratedQuestion {
    pub id: String,
    pub skill_id: i64,
    pub section: String,
    pub interaction: String,
    pub stimulus: Option<Value>,
    pub prompt: String,
    pub choices: Option<Vec<String>>,
    pub correct_response: Option<Vec<String>>,
    pub cefr_target: Option<String>,
    pub difficulty_score: Option<i64>,
    pub passage_id: Option<String>,
    pub metadata: Option<Value>,
    pub created_at: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct QuizGenerateResponse {
    pub questions: Vec<GeneratedQuestion>,
}

#[derive(Debug, Deserialize)]
pub struct QuestionCreateRequest {
    pub skill_id: i64,
    pub section: String,
    pub interaction: String,
    pub prompt: String,
    pub choices: Option<Vec<String>>,
    pub correct_response: Option<Vec<String>>,
    pub cefr_target: Option<String>,
    pub difficulty_score: Option<i64>,
    pub passage_id: Option<String>,
    pub stimulus: Option<String>,
    pub metadata: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct QuestionUpdateRequest {
    pub skill_id: Option<i64>,
    pub section: Option<String>,
    pub interaction: Option<String>,
    pub prompt: Option<String>,
    pub choices: Option<Vec<String>>,
    pub correct_response: Option<Vec<String>>,
    pub cefr_target: Option<String>,
    pub difficulty_score: Option<i64>,
    pub passage_id: Option<String>,
    pub stimulus: Option<String>,
    pub metadata: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct PassageCreateRequest {
    pub topic: Option<String>,
    pub content: String,
    pub source: Option<String>,
    pub difficulty: Option<String>,
}
